be/dtm0_log.h

```diff
@@ -0,0 +1,522 @@
+ *    m0_be_dtm0_log during mkfs procedure.
+ *
+ * 2. When a distributed transaction(dtx) is successfully executed on a participant,
+ *    the participant will add a new log record to DTM0 log. The state of the
+ *    participant will be set to M0_DTPS_EXECUTED in the record. The states of the
+ *    other participants will be set to M0_DTPS_INPROGRESS in the record.
+ *
+ * 3. When a distributed transaction(dtx) becomes persistent on a particular
+ *    participant, the state of the distributed transaction(dtx) for this
+ *    participant will be updated to M0_DTPS_PERSISTENT.
+ *
+ * 4. When distributed transaction(dtx) becomes persistent on a participant,
+ *    this participant sends a persistent notice to rest of the participants
+ *    and to the originator indicating that the given distributed transaction(dtx)
+ *    is now persistent on its store. Upon receiving a persistent notice, the
+ *    recipient will update the state of the distributed transaction(dtx) for the
+ *    sender as M0_DTPS_PERSISTENT in its own DTM0 log.
+ *
+ *    There is one special case where a participant gets a persistent notice from
+ *    another participant but the corresponding request and reply haven't been
+ *    received/processed yet. In this case, the DTM service should add a special
+ *    log record: a log record with empty payload (NULL). The users of DTM0 log
+ *    should be ready to encounter such a log entry, and handle it appropriately.
+ */
+
+/**
+ * @defgroup DTM0Internals DTM0 implementation Internals
+ * @ingroup DTM0
+ *
+ * @section This section describes the dtm0 log related enumerations and structures.
+ *
+ * @{
+ */
+
+/**
+ * @b  m0_be_dtm0_log_credit_op enum
+ *
+ * DTM0 persistent log is implemented on top of be modules and uses be_list for
+ * storing the log records. m0_be_dtm0_log_credit_op is an enumeration of the
+ * operation codes that can be specified to obtain credits for performing the
+ * corresponding operation related to dtm0_log.
+ */
+enum m0_be_dtm0_log_credit_op {
+	M0_DTML_CREATE,		/**< m0_be_dtm0_log_create() */
+	M0_DTML_SENT,		/**< m0_be_dtm0_log_update() */
+	M0_DTML_EXECUTED,	/**< m0_be_dtm0_log_update() */
+	M0_DTML_PERSISTENT,	/**< m0_be_dtm0_log_update() */
+	M0_DTML_PRUNE,		/**< m0_be_dtm0_log_prune()  */
+	M0_DTML_REDO            /**< m0_be_dtm0_log_update() */
+};
+
+/**
+ * @b  m0_be_dtm0_log_rec structure
+ *
+ * A DTM0 log record is represented by m0_dtm0_log_rec structure. The important
+ * fields in this structure are:
+ * - dlr_dtx: This stores dtx information related to dtm0 client.
+ * - dlr_txd: This stores the states of the participants.
+ * - dlr_payload: This stores the original cas_op (cas request).
+ */
+
+struct m0_dtm0_log_rec {
+	struct m0_dtm0_dtx     dlr_dtx;
+	struct m0_dtm0_tx_desc dlr_txd;
+	uint64_t               dlr_magic;
+	union  {
+		struct m0_be_list_link dlr_link;    /**< link into m0_be_dtm0_log::tlist */
+		struct m0_tlink        dlr_tlink;
```

Comment:
Please add comments (though this is easy to guess/understand).
dlr_link - used when this is a persistent log that is going to stored in BE.
dlr_tlink - used when this is a volatile log that is only in-memory.

Reply:
How do we know if a record is persistent or volatile?
Because a record is linked to different lists (m0_be_list_link or m0_tlink)?

Reply:
Done.